compiler/plugins/target/ROCM/builtins/ukernel/argmax_ukernel.c

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <float.h>
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+
+extern "C" __device__ __attribute__((const)) half __ockl_wfred_max_f16(half);
+extern "C" __device__ __attribute__((const)) float __ockl_wfred_max_f32(float);
+extern "C" __device__ __attribute__((const))
+int64_t __ockl_wfred_min_i64(int64_t);
+extern "C" __device__ __attribute__((const))
+int32_t __ockl_wfred_min_i32(int32_t);
+
+extern "C" __device__ void
+__iree_uk_rocm_argmax_F32I32(float *inputBuffer, size_t input_offset,
+                             int32_t *outputBuffer, size_t output_offset,
+                             size_t reductionSize, uint32_t flag) {
+  uint laneID = __builtin_amdgcn_workitem_id_x();
+  // Set identity value to handle problem non divisible by subgroupSize.
+  float laneMax =
+      laneID >= reductionSize ? -FLT_MAX : inputBuffer[input_offset + laneID];
+  int32_t laneResult = laneID;
+
+  // NOTE: On F32 kernels with clang, reductionSize/blockDim.x has numerical
+  // inaccuracy.
+  uint numBatches = reductionSize / warpSize + 1;
```

Comment:
ceildiv? Isn't this always running an additional step with `-FLT_MAX` when `reductionSize` is a multiplier of `warpSize`?

Reply:
Good question, the for loop below this actually starts from 1, because we want to use the induction variable `i` as the offset. So we set the numBatches to be ceilDiv.

Reply:
My understanding of `i` starting with `1` is because you've already processed one element in the above. For example, when we have `reductionSize = warpSize = 32`, the following loop should not run at all? At the moment `numBatches` would be 2 and we still have one additional iteration doing `-FLT_MAX`? I think we should have `numBatches = (redutionSize + warpSize - 1) / warpSize`?

Reply:
done, ahhh you're right! good catch, thanks for this! :)